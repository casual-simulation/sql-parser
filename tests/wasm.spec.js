import { initSync, parse_sql, SQLVisitor, visit } from '../pkg/sql_parser_wasm';
import { readFile } from 'node:fs/promises';
import {jest} from '@jest/globals';

beforeAll(async () => {
    let wasmPath = new URL(import.meta.resolve('../pkg/sql_parser_wasm_bg.wasm')).pathname;
    if (process.platform === 'win32' && wasmPath.startsWith('/')) {
        wasmPath = wasmPath.slice(1);
    }
    const bytes = await readFile(wasmPath);
    initSync({
        module: bytes
    });
});

it('should be able to parse some SQL', () => {
    const result = parse_sql("generic", "SELECT * FROM users;");

    expect(result).toMatchSnapshot();
});

it('should be able to parse insert SQL', () => {
    const result = parse_sql("generic", "INSERT INTO users (id, name) VALUES (1, 'Alice'), (2, 'Bob');");

    expect(result).toMatchSnapshot();
});

it('should be able to parse update SQL', () => {
    const result = parse_sql("generic", "UPDATE users SET name = 'Charlie' WHERE id = 1;");

    expect(result).toMatchSnapshot();
});

it('should be able to parse delete SQL', () => {
    const result = parse_sql("generic", "DELETE FROM users WHERE id = 2;");

    expect(result).toMatchSnapshot();
});

it('should support parameterized queries', () => {
    const result = parse_sql("generic", "SELECT * FROM users WHERE id = ?;");

    expect(result).toMatchSnapshot();
});

it('should be able to parse joins', () => {
    const result = parse_sql("generic", "SELECT * FROM users INNER JOIN orders ON users.id = orders.user_id;");
    expect(result).toMatchSnapshot();
});

it('should throw errors when parsing invalid SQL', () => {
    expect(() => parse_sql("generic", "SELEC * FROM users;")).toThrow("sql parser error: Expected: an SQL statement, found: SELEC at Line: 1, Column: 1");
});

describe('visitor', () => {
    let visitor;
    let pre_visit_query;
    let post_visit_query;
    let pre_visit_relation;
    let post_visit_relation;
    let pre_visit_table_factor;
    let post_visit_table_factor;
    let pre_visit_expr;
    let post_visit_expr;
    let pre_visit_statement;
    let post_visit_statement;
    let pre_visit_value;
    let post_visit_value;

    beforeEach(() => {
        pre_visit_query = jest.fn();
        post_visit_query = jest.fn();
        pre_visit_relation = jest.fn();
        post_visit_relation = jest.fn();
        pre_visit_table_factor = jest.fn();
        post_visit_table_factor = jest.fn();
        pre_visit_expr = jest.fn();
        post_visit_expr = jest.fn();
        pre_visit_statement = jest.fn();
        post_visit_statement = jest.fn();
        pre_visit_value = jest.fn();
        post_visit_value = jest.fn();

        visitor = new SQLVisitor({
            pre_visit_query,
            post_visit_query,
            pre_visit_relation,
            post_visit_relation,
            pre_visit_table_factor,
            post_visit_table_factor,
            pre_visit_expr,
            post_visit_expr,
            pre_visit_statement,
            post_visit_statement,
            pre_visit_value,
            post_visit_value,
        });
    });

    it('should call visitor methods when visiting a statement', () => {
        const result = parse_sql("generic", "SELECT * FROM users;");
        expect(result.length).toBe(1);
        const statement = result[0];

        visit(visitor, statement);

        expect(pre_visit_statement).toHaveBeenCalledTimes(1);
        expect(post_visit_statement).toHaveBeenCalledTimes(1);
        expect(pre_visit_query).toHaveBeenCalledTimes(1);
        expect(post_visit_query).toHaveBeenCalledTimes(1);
        expect(pre_visit_relation).toHaveBeenCalledTimes(1);
        expect(post_visit_relation).toHaveBeenCalledTimes(1);
        expect(pre_visit_table_factor).toHaveBeenCalledTimes(1);
        expect(post_visit_table_factor).toHaveBeenCalledTimes(1);
        expect(pre_visit_expr).toHaveBeenCalledTimes(0);
        expect(post_visit_expr).toHaveBeenCalledTimes(0);
    });

    it('should call the statement visitors', () => {
        const result = parse_sql("generic", "SELECT * FROM users;");
        expect(result.length).toBe(1);
        const statement = result[0];

        visit(visitor, statement);

        expect(pre_visit_statement).toHaveBeenCalledTimes(1);
        expect(post_visit_statement).toHaveBeenCalledTimes(1);
        expect(pre_visit_statement).toHaveBeenCalledWith(statement);
        expect(post_visit_statement).toHaveBeenCalledWith(statement);
    });

    it('should call the query visitors', () => {
        const result = parse_sql("generic", "SELECT * FROM users;");
        expect(result.length).toBe(1);

        const statement = result[0];
        
        visit(visitor, statement);

        expect(pre_visit_query).toHaveBeenCalledTimes(1);
        expect(post_visit_query).toHaveBeenCalledTimes(1);
        expect(pre_visit_query).toHaveBeenCalledWith(statement.Query);
        expect(post_visit_query).toHaveBeenCalledWith(statement.Query);
    });

    it('should call the table factor visitors', () => {
        const result = parse_sql("generic", "SELECT * FROM users INNER JOIN orders ON users.id = orders.user_id;");
        expect(result.length).toBe(1);

        const statement = result[0];
        
        visit(visitor, statement);

        expect(pre_visit_table_factor).toHaveBeenCalledTimes(2);
        expect(post_visit_table_factor).toHaveBeenCalledTimes(2);
        expect(pre_visit_table_factor).toHaveBeenCalledWith({
            Table: {
                index_hints: [],
                partitions: [],
                with_hints: [],
                with_ordinality: false,
                name: [
                    {
                        Identifier: {
                            value: 'users',
                            span: expect.any(Object)
                        }
                    }
                ]
            }
        });
        expect(pre_visit_table_factor).toHaveBeenCalledWith({
            Table: {
                index_hints: [],
                partitions: [],
                with_hints: [],
                with_ordinality: false,
                name: [
                    {
                        Identifier: {
                            value: 'orders',
                            span: expect.any(Object)
                        }
                    }
                ]
            }
        });
        expect(post_visit_table_factor).toHaveBeenCalledWith({
            Table: {
                index_hints: [],
                partitions: [],
                with_hints: [],
                with_ordinality: false,
                name: [
                    {
                        Identifier: {
                            value: 'users',
                            span: expect.any(Object)
                        }
                    }
                ]
            }
        });
        expect(post_visit_table_factor).toHaveBeenCalledWith({
            Table: {
                index_hints: [],
                partitions: [],
                with_hints: [],
                with_ordinality: false,
                name: [
                    {
                        Identifier: {
                            value: 'orders',
                            span: expect.any(Object)
                        }
                    }
                ]
            }
        });
    });

    it('should call the relation visitors', () => {
        const result = parse_sql("generic", "SELECT * FROM users INNER JOIN orders ON users.id = orders.user_id;");
        expect(result.length).toBe(1);

        const statement = result[0];
        
        visit(visitor, statement);

        expect(pre_visit_relation).toHaveBeenCalledTimes(2);
        expect(post_visit_relation).toHaveBeenCalledTimes(2);
        expect(pre_visit_relation).toHaveBeenCalledWith([{
            Identifier: {
                value: 'users',
                span: expect.any(Object)
            }
        }]);
        expect(pre_visit_relation).toHaveBeenCalledWith([{
            Identifier: {
                value: 'orders',
                span: expect.any(Object)
            }
        }]);
        expect(post_visit_relation).toHaveBeenCalledWith([{
            Identifier: {
                value: 'users',
                span: expect.any(Object)
            }
        }]);
        expect(post_visit_relation).toHaveBeenCalledWith([{
            Identifier: {
                value: 'orders',
                span: expect.any(Object)
            }
        }]);
    });

    it('should call the value visitors', () => {
        const result = parse_sql("generic", "INSERT INTO users (id, name) VALUES (1, 'Alice'), (2, 'Bob');");
        expect(result.length).toBe(1);

        const statement = result[0];

        visit(visitor, statement);

        expect(pre_visit_value).toHaveBeenCalledTimes(4);
        expect(post_visit_value).toHaveBeenCalledTimes(4);
        expect(pre_visit_value).toHaveBeenCalledWith({ Number: [ "1", false ] });
        expect(pre_visit_value).toHaveBeenCalledWith({ Number: [ "2", false ] });
        expect(pre_visit_value).toHaveBeenCalledWith({ SingleQuotedString: "Alice" });
        expect(pre_visit_value).toHaveBeenCalledWith({ SingleQuotedString: "Bob" });
    });

    it('should call the expr visitors', () => {
        const result = parse_sql("generic", "SELECT * FROM users WHERE id = 1 AND name = 'Alice';");
        expect(result.length).toBe(1);

        const statement = result[0];
        visit(visitor, statement);

        expect(pre_visit_expr).toHaveBeenCalledTimes(7);
        expect(post_visit_expr).toHaveBeenCalledTimes(7);
    });
});