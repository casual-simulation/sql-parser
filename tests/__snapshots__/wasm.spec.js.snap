// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should be able to parse delete SQL 1`] = `
[
  {
    "Delete": {
      "from": {
        "WithFromKeyword": [
          {
            "joins": [],
            "relation": {
              "Table": {
                "alias": undefined,
                "args": undefined,
                "index_hints": [],
                "json_path": undefined,
                "name": [
                  {
                    "Identifier": {
                      "quote_style": undefined,
                      "span": {
                        "end": {
                          "column": 18,
                          "line": 1,
                        },
                        "start": {
                          "column": 13,
                          "line": 1,
                        },
                      },
                      "value": "users",
                    },
                  },
                ],
                "partitions": [],
                "sample": undefined,
                "version": undefined,
                "with_hints": [],
                "with_ordinality": false,
              },
            },
          },
        ],
      },
      "limit": undefined,
      "order_by": [],
      "returning": undefined,
      "selection": {
        "BinaryOp": {
          "left": {
            "Identifier": {
              "quote_style": undefined,
              "span": {
                "end": {
                  "column": 27,
                  "line": 1,
                },
                "start": {
                  "column": 25,
                  "line": 1,
                },
              },
              "value": "id",
            },
          },
          "op": "Eq",
          "right": {
            "Value": {
              "span": {
                "end": {
                  "column": 31,
                  "line": 1,
                },
                "start": {
                  "column": 30,
                  "line": 1,
                },
              },
              "value": {
                "Number": [
                  "2",
                  false,
                ],
              },
            },
          },
        },
      },
      "tables": [],
      "using": undefined,
    },
  },
]
`;

exports[`should be able to parse insert SQL 1`] = `
[
  {
    "Insert": {
      "after_columns": [],
      "assignments": [],
      "columns": [
        {
          "quote_style": undefined,
          "span": {
            "end": {
              "column": 22,
              "line": 1,
            },
            "start": {
              "column": 20,
              "line": 1,
            },
          },
          "value": "id",
        },
        {
          "quote_style": undefined,
          "span": {
            "end": {
              "column": 28,
              "line": 1,
            },
            "start": {
              "column": 24,
              "line": 1,
            },
          },
          "value": "name",
        },
      ],
      "format_clause": undefined,
      "has_table_keyword": false,
      "ignore": false,
      "insert_alias": undefined,
      "into": true,
      "on": undefined,
      "or": undefined,
      "overwrite": false,
      "partitioned": undefined,
      "priority": undefined,
      "replace_into": false,
      "returning": undefined,
      "settings": undefined,
      "source": {
        "body": {
          "Values": {
            "explicit_row": false,
            "rows": [
              [
                {
                  "Value": {
                    "span": {
                      "end": {
                        "column": 39,
                        "line": 1,
                      },
                      "start": {
                        "column": 38,
                        "line": 1,
                      },
                    },
                    "value": {
                      "Number": [
                        "1",
                        false,
                      ],
                    },
                  },
                },
                {
                  "Value": {
                    "span": {
                      "end": {
                        "column": 48,
                        "line": 1,
                      },
                      "start": {
                        "column": 41,
                        "line": 1,
                      },
                    },
                    "value": {
                      "SingleQuotedString": "Alice",
                    },
                  },
                },
              ],
              [
                {
                  "Value": {
                    "span": {
                      "end": {
                        "column": 53,
                        "line": 1,
                      },
                      "start": {
                        "column": 52,
                        "line": 1,
                      },
                    },
                    "value": {
                      "Number": [
                        "2",
                        false,
                      ],
                    },
                  },
                },
                {
                  "Value": {
                    "span": {
                      "end": {
                        "column": 60,
                        "line": 1,
                      },
                      "start": {
                        "column": 55,
                        "line": 1,
                      },
                    },
                    "value": {
                      "SingleQuotedString": "Bob",
                    },
                  },
                },
              ],
            ],
          },
        },
        "fetch": undefined,
        "for_clause": undefined,
        "format_clause": undefined,
        "limit_clause": undefined,
        "locks": [],
        "order_by": undefined,
        "pipe_operators": [],
        "settings": undefined,
        "with": undefined,
      },
      "table": {
        "TableName": [
          {
            "Identifier": {
              "quote_style": undefined,
              "span": {
                "end": {
                  "column": 18,
                  "line": 1,
                },
                "start": {
                  "column": 13,
                  "line": 1,
                },
              },
              "value": "users",
            },
          },
        ],
      },
      "table_alias": undefined,
    },
  },
]
`;

exports[`should be able to parse joins 1`] = `
[
  {
    "Query": {
      "body": {
        "Select": {
          "cluster_by": [],
          "connect_by": undefined,
          "distinct": undefined,
          "distribute_by": [],
          "exclude": undefined,
          "flavor": "Standard",
          "from": [
            {
              "joins": [
                {
                  "global": false,
                  "join_operator": {
                    "Inner": {
                      "On": {
                        "BinaryOp": {
                          "left": {
                            "CompoundIdentifier": [
                              {
                                "quote_style": undefined,
                                "span": {
                                  "end": {
                                    "column": 47,
                                    "line": 1,
                                  },
                                  "start": {
                                    "column": 42,
                                    "line": 1,
                                  },
                                },
                                "value": "users",
                              },
                              {
                                "quote_style": undefined,
                                "span": {
                                  "end": {
                                    "column": 50,
                                    "line": 1,
                                  },
                                  "start": {
                                    "column": 48,
                                    "line": 1,
                                  },
                                },
                                "value": "id",
                              },
                            ],
                          },
                          "op": "Eq",
                          "right": {
                            "CompoundIdentifier": [
                              {
                                "quote_style": undefined,
                                "span": {
                                  "end": {
                                    "column": 59,
                                    "line": 1,
                                  },
                                  "start": {
                                    "column": 53,
                                    "line": 1,
                                  },
                                },
                                "value": "orders",
                              },
                              {
                                "quote_style": undefined,
                                "span": {
                                  "end": {
                                    "column": 67,
                                    "line": 1,
                                  },
                                  "start": {
                                    "column": 60,
                                    "line": 1,
                                  },
                                },
                                "value": "user_id",
                              },
                            ],
                          },
                        },
                      },
                    },
                  },
                  "relation": {
                    "Table": {
                      "alias": undefined,
                      "args": undefined,
                      "index_hints": [],
                      "json_path": undefined,
                      "name": [
                        {
                          "Identifier": {
                            "quote_style": undefined,
                            "span": {
                              "end": {
                                "column": 38,
                                "line": 1,
                              },
                              "start": {
                                "column": 32,
                                "line": 1,
                              },
                            },
                            "value": "orders",
                          },
                        },
                      ],
                      "partitions": [],
                      "sample": undefined,
                      "version": undefined,
                      "with_hints": [],
                      "with_ordinality": false,
                    },
                  },
                },
              ],
              "relation": {
                "Table": {
                  "alias": undefined,
                  "args": undefined,
                  "index_hints": [],
                  "json_path": undefined,
                  "name": [
                    {
                      "Identifier": {
                        "quote_style": undefined,
                        "span": {
                          "end": {
                            "column": 20,
                            "line": 1,
                          },
                          "start": {
                            "column": 15,
                            "line": 1,
                          },
                        },
                        "value": "users",
                      },
                    },
                  ],
                  "partitions": [],
                  "sample": undefined,
                  "version": undefined,
                  "with_hints": [],
                  "with_ordinality": false,
                },
              },
            },
          ],
          "group_by": {
            "Expressions": [
              [],
              [],
            ],
          },
          "having": undefined,
          "into": undefined,
          "lateral_views": [],
          "named_window": [],
          "prewhere": undefined,
          "projection": [
            {
              "Wildcard": {
                "opt_except": undefined,
                "opt_exclude": undefined,
                "opt_ilike": undefined,
                "opt_rename": undefined,
                "opt_replace": undefined,
                "wildcard_token": {
                  "span": {
                    "end": {
                      "column": 9,
                      "line": 1,
                    },
                    "start": {
                      "column": 8,
                      "line": 1,
                    },
                  },
                  "token": "Mul",
                },
              },
            },
          ],
          "qualify": undefined,
          "select_token": {
            "span": {
              "end": {
                "column": 7,
                "line": 1,
              },
              "start": {
                "column": 1,
                "line": 1,
              },
            },
            "token": {
              "Word": {
                "keyword": "SELECT",
                "quote_style": undefined,
                "value": "SELECT",
              },
            },
          },
          "selection": undefined,
          "sort_by": [],
          "top": undefined,
          "top_before_distinct": false,
          "value_table_mode": undefined,
          "window_before_qualify": false,
        },
      },
      "fetch": undefined,
      "for_clause": undefined,
      "format_clause": undefined,
      "limit_clause": undefined,
      "locks": [],
      "order_by": undefined,
      "pipe_operators": [],
      "settings": undefined,
      "with": undefined,
    },
  },
]
`;

exports[`should be able to parse some SQL 1`] = `
[
  {
    "Query": {
      "body": {
        "Select": {
          "cluster_by": [],
          "connect_by": undefined,
          "distinct": undefined,
          "distribute_by": [],
          "exclude": undefined,
          "flavor": "Standard",
          "from": [
            {
              "joins": [],
              "relation": {
                "Table": {
                  "alias": undefined,
                  "args": undefined,
                  "index_hints": [],
                  "json_path": undefined,
                  "name": [
                    {
                      "Identifier": {
                        "quote_style": undefined,
                        "span": {
                          "end": {
                            "column": 20,
                            "line": 1,
                          },
                          "start": {
                            "column": 15,
                            "line": 1,
                          },
                        },
                        "value": "users",
                      },
                    },
                  ],
                  "partitions": [],
                  "sample": undefined,
                  "version": undefined,
                  "with_hints": [],
                  "with_ordinality": false,
                },
              },
            },
          ],
          "group_by": {
            "Expressions": [
              [],
              [],
            ],
          },
          "having": undefined,
          "into": undefined,
          "lateral_views": [],
          "named_window": [],
          "prewhere": undefined,
          "projection": [
            {
              "Wildcard": {
                "opt_except": undefined,
                "opt_exclude": undefined,
                "opt_ilike": undefined,
                "opt_rename": undefined,
                "opt_replace": undefined,
                "wildcard_token": {
                  "span": {
                    "end": {
                      "column": 9,
                      "line": 1,
                    },
                    "start": {
                      "column": 8,
                      "line": 1,
                    },
                  },
                  "token": "Mul",
                },
              },
            },
          ],
          "qualify": undefined,
          "select_token": {
            "span": {
              "end": {
                "column": 7,
                "line": 1,
              },
              "start": {
                "column": 1,
                "line": 1,
              },
            },
            "token": {
              "Word": {
                "keyword": "SELECT",
                "quote_style": undefined,
                "value": "SELECT",
              },
            },
          },
          "selection": undefined,
          "sort_by": [],
          "top": undefined,
          "top_before_distinct": false,
          "value_table_mode": undefined,
          "window_before_qualify": false,
        },
      },
      "fetch": undefined,
      "for_clause": undefined,
      "format_clause": undefined,
      "limit_clause": undefined,
      "locks": [],
      "order_by": undefined,
      "pipe_operators": [],
      "settings": undefined,
      "with": undefined,
    },
  },
]
`;

exports[`should be able to parse update SQL 1`] = `
[
  {
    "Update": {
      "assignments": [
        {
          "target": {
            "ColumnName": [
              {
                "Identifier": {
                  "quote_style": undefined,
                  "span": {
                    "end": {
                      "column": 22,
                      "line": 1,
                    },
                    "start": {
                      "column": 18,
                      "line": 1,
                    },
                  },
                  "value": "name",
                },
              },
            ],
          },
          "value": {
            "Value": {
              "span": {
                "end": {
                  "column": 34,
                  "line": 1,
                },
                "start": {
                  "column": 25,
                  "line": 1,
                },
              },
              "value": {
                "SingleQuotedString": "Charlie",
              },
            },
          },
        },
      ],
      "from": undefined,
      "limit": undefined,
      "or": undefined,
      "returning": undefined,
      "selection": {
        "BinaryOp": {
          "left": {
            "Identifier": {
              "quote_style": undefined,
              "span": {
                "end": {
                  "column": 43,
                  "line": 1,
                },
                "start": {
                  "column": 41,
                  "line": 1,
                },
              },
              "value": "id",
            },
          },
          "op": "Eq",
          "right": {
            "Value": {
              "span": {
                "end": {
                  "column": 47,
                  "line": 1,
                },
                "start": {
                  "column": 46,
                  "line": 1,
                },
              },
              "value": {
                "Number": [
                  "1",
                  false,
                ],
              },
            },
          },
        },
      },
      "table": {
        "joins": [],
        "relation": {
          "Table": {
            "alias": undefined,
            "args": undefined,
            "index_hints": [],
            "json_path": undefined,
            "name": [
              {
                "Identifier": {
                  "quote_style": undefined,
                  "span": {
                    "end": {
                      "column": 13,
                      "line": 1,
                    },
                    "start": {
                      "column": 8,
                      "line": 1,
                    },
                  },
                  "value": "users",
                },
              },
            ],
            "partitions": [],
            "sample": undefined,
            "version": undefined,
            "with_hints": [],
            "with_ordinality": false,
          },
        },
      },
    },
  },
]
`;

exports[`should support parameterized queries 1`] = `
[
  {
    "Query": {
      "body": {
        "Select": {
          "cluster_by": [],
          "connect_by": undefined,
          "distinct": undefined,
          "distribute_by": [],
          "exclude": undefined,
          "flavor": "Standard",
          "from": [
            {
              "joins": [],
              "relation": {
                "Table": {
                  "alias": undefined,
                  "args": undefined,
                  "index_hints": [],
                  "json_path": undefined,
                  "name": [
                    {
                      "Identifier": {
                        "quote_style": undefined,
                        "span": {
                          "end": {
                            "column": 20,
                            "line": 1,
                          },
                          "start": {
                            "column": 15,
                            "line": 1,
                          },
                        },
                        "value": "users",
                      },
                    },
                  ],
                  "partitions": [],
                  "sample": undefined,
                  "version": undefined,
                  "with_hints": [],
                  "with_ordinality": false,
                },
              },
            },
          ],
          "group_by": {
            "Expressions": [
              [],
              [],
            ],
          },
          "having": undefined,
          "into": undefined,
          "lateral_views": [],
          "named_window": [],
          "prewhere": undefined,
          "projection": [
            {
              "Wildcard": {
                "opt_except": undefined,
                "opt_exclude": undefined,
                "opt_ilike": undefined,
                "opt_rename": undefined,
                "opt_replace": undefined,
                "wildcard_token": {
                  "span": {
                    "end": {
                      "column": 9,
                      "line": 1,
                    },
                    "start": {
                      "column": 8,
                      "line": 1,
                    },
                  },
                  "token": "Mul",
                },
              },
            },
          ],
          "qualify": undefined,
          "select_token": {
            "span": {
              "end": {
                "column": 7,
                "line": 1,
              },
              "start": {
                "column": 1,
                "line": 1,
              },
            },
            "token": {
              "Word": {
                "keyword": "SELECT",
                "quote_style": undefined,
                "value": "SELECT",
              },
            },
          },
          "selection": {
            "BinaryOp": {
              "left": {
                "Identifier": {
                  "quote_style": undefined,
                  "span": {
                    "end": {
                      "column": 29,
                      "line": 1,
                    },
                    "start": {
                      "column": 27,
                      "line": 1,
                    },
                  },
                  "value": "id",
                },
              },
              "op": "Eq",
              "right": {
                "Value": {
                  "span": {
                    "end": {
                      "column": 33,
                      "line": 1,
                    },
                    "start": {
                      "column": 32,
                      "line": 1,
                    },
                  },
                  "value": {
                    "Placeholder": "?",
                  },
                },
              },
            },
          },
          "sort_by": [],
          "top": undefined,
          "top_before_distinct": false,
          "value_table_mode": undefined,
          "window_before_qualify": false,
        },
      },
      "fetch": undefined,
      "for_clause": undefined,
      "format_clause": undefined,
      "limit_clause": undefined,
      "locks": [],
      "order_by": undefined,
      "pipe_operators": [],
      "settings": undefined,
      "with": undefined,
    },
  },
]
`;
